### PR TITLE
Add '--logging-format' flag to kube-proxy

### LINKED ISSUE #105165: Introduce JSON logging format for kube-proxy
#### What would you like to be added:

Structured logging support in k8s allows specifying logging in JSON. Many k8s components like apiServer, scheduler have support for json. This issue tracks adding JSON logging support for kube-proxy.  

#### Breakdown of the task

To get json logging support, changes have to be made in kube-proxy and Kind e2e script (for Prow/CI runs) which uses KubeProxyConfiguration 

1. Introduce logging-format (--logging-format string  ) option in kube-proxy. Currently there is none. ([link](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-proxy/))
2. Introduce an option in [KubeProxyConfiguration](https://github.com/kubernetes/kube-proxy/blob/v0.22.1/config/v1alpha1/types.go#L104) so that kubeadm/kind can specify logging-format. Whether this "configuration option" is something generic or specific to logging needs input from a kube-proxy domain expert.
3. Update Kind to make use of the above 2 and specify the logging format similar to how we do for control-plane components. 
4. Prow/test-infra CI runs will have json logging enabled for kube-proxy.

#### Why is this needed:
To get the benefits of structured logging.

/sig network
/cc @serathius 
/cc @BenTheElder

### PR DESCRIPTION
**Pull request by cyclinder**
#### What type of PR is this?

/kind feature
#### What this PR does / why we need it:

Add '--logging-format' flag to kube-proxy

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/105165

#### Does this PR introduce a user-facing change?

```release-note
kube-proxy: add '--logging-format' flag to support structured logging
```

### CODE REVIEW

File: cmd/kube-proxy/app/server.go
Comment (danwinship) [inline cmd/kube-proxy/app/server.go:24]: sorted incorrectly; "namespaced" imports are kept separate from standard library imports

(likewise in other files)

File: cmd/kube-proxy/app/server.go
Comment (danwinship) [inline cmd/kube-proxy/app/server.go:471]: this comment seems to have been copied from the kubelet sources but it doesn't actually seem to apply here?

I'm not sure this is the right place for the InitLogs, etc, code either. How did you decide to put it here as opposed to anywhere else?

File: cmd/kube-proxy/app/server_test.go
Comment (danwinship) [inline cmd/kube-proxy/app/server_test.go:292]: This seems to be removing the check that the rest of `config` is correct

File: pkg/proxy/apis/config/types.go
Comment (danwinship) [inline pkg/proxy/apis/config/types.go:193]:
```suggestion
	// Refer to [Logs Options](https://github.com/kubernetes/component-base/blob/master/logs/options.go) for more information.
```

File: pkg/proxy/apis/config/scheme/testdata/KubeProxyConfiguration/after/v1alpha1.yaml
Comment (danwinship) [inline pkg/proxy/apis/config/scheme/testdata/KubeProxyConfiguration/after/v1alpha1.yaml:41]: this seems like a bug; it should serialize to "`5s`", not "`5000000000`"

File: pkg/proxy/apis/config/validation/validation_test.go
Comment (danwinship) [inline pkg/proxy/apis/config/validation/validation_test.go:516]: this test would probably be better with an _explicit_ bad value

**Review by danwinship — CHANGES_REQUESTED** [review]

File: cmd/kube-proxy/app/server.go
Comment (cyclinder) [reply cmd/kube-proxy/app/server.go:471]: Yes, copied here from the kubelet sources. `InitLogs` should be called after parsing flags and before cliflag.PrintFlags. 

Refer to:https://github.com/kubernetes/kubernetes/blob/b32941f60d3ba789965bbf32f4f5bf756feab075/staging/src/k8s.io/component-base/logs/logs.go#L149
https://github.com/kubernetes/kubernetes/blob/b32941f60d3ba789965bbf32f4f5bf756feab075/cmd/kube-proxy/app/server.go#L468

I went through the startup source code of kube-apiserver, kube-controller-manager, kube-scheduler, and kubelet.

Let's take kube-apiserver as an example:

- New log options: https://github.com/kubernetes/kubernetes/blob/b32941f60d3ba789965bbf32f4f5bf756feab075/cmd/kube-apiserver/app/options/options.go#L107

- Add log flag: https://github.com/kubernetes/kubernetes/blob/b32941f60d3ba789965bbf32f4f5bf756feab075/cmd/kube-apiserver/app/options/options.go#L156

- ValidateAndApply log options: https://github.com/kubernetes/kubernetes/blob/b32941f60d3ba789965bbf32f4f5bf756feab075/cmd/kube-apiserver/app/server.go#L119

kube-proxy is similar to kubelet but different from the other components. so here is the same as kubelet. Please see https://github.com/kubernetes/kubernetes/pull/105797/files#diff-3a14a9838132c6668f4d100c26afa05b471b149c469edfece9b84fcd058c8a4b


Maybe this needs to be ack by sig/instrumentation reviewer

/cc @pohly @liggitt

File: cmd/kube-proxy/app/server_test.go
Comment (cyclinder) [reply cmd/kube-proxy/app/server_test.go:292]: oh thanks, it's my test code here

File: pkg/proxy/apis/config/types.go
Comment (cyclinder) [reply pkg/proxy/apis/config/types.go:193]: Updated.

File: cmd/kube-proxy/app/server.go
Comment (cyclinder) [reply cmd/kube-proxy/app/server.go:24]: Updated.

File: pkg/proxy/apis/config/scheme/testdata/KubeProxyConfiguration/after/v1alpha1.yaml
Comment (cyclinder) [reply pkg/proxy/apis/config/scheme/testdata/KubeProxyConfiguration/after/v1alpha1.yaml:41]: After serialize, the default values for logging are indeed these, see https://github.com/kubernetes/kubernetes/blob/b248684334c50027feb20cf42bcc8c48e4f9ede4/pkg/kubelet/apis/config/scheme/testdata/KubeletConfiguration/after/v1beta1.yaml#L54-L59

File: pkg/proxy/apis/config/validation/validation_test.go
Comment (cyclinder) [reply pkg/proxy/apis/config/validation/validation_test.go:516]: Updated.

File: pkg/proxy/apis/config/scheme/testdata/KubeProxyConfiguration/after/v1alpha1.yaml
Comment (danwinship) [reply pkg/proxy/apis/config/scheme/testdata/KubeProxyConfiguration/after/v1alpha1.yaml:41]: This seems like it's a bug in the shared component-base code then. Ah, known bug, couldn't be fixed because of API concerns. Probably to be fixed with a new field later. #107618

File: cmd/kube-proxy/app/server.go
Comment (danwinship) [reply cmd/kube-proxy/app/server.go:471]: The config file is not parsed until a few lines later (from `opts.Complete`)

File: pkg/proxy/apis/config/scheme/testdata/KubeProxyConfiguration/after/v1alpha1.yaml
Comment (liggitt) [reply pkg/proxy/apis/config/scheme/testdata/KubeProxyConfiguration/after/v1alpha1.yaml:41]: before we expand to using this in a new component, can we fix this up... a type that accepts reading either int time.Duration or string metav1.Duration would make this more usable without breaking reading existing config files

sketch in the last commit of https://github.com/liggitt/kubernetes/commits/flushType

File: cmd/kube-proxy/app/server.go
Comment (liggitt) [inline cmd/kube-proxy/app/server.go:473]: I don't think this is the right place to do this... until you run `Complete()` below, you haven't actually loaded the config file, so no log options in the config will be read yet here, and no feature gates values will have been loaded from the config file either

moving this after `Complete()` would mean the config would be loaded and feature gates from the config set

Comment (cyclinder) [reply cmd/kube-proxy/app/server.go]: Thanks, updated. 

I build a kube-proxy binary in my local to make a test, I passed `--logging-format json` and started it, but the start fails and returns："initialize logging: format: Invalid value: \"json\": Unsupported log format"

Is there anywhere else I missed?

```shell
➜  kubernetes git:(loggin_format) ✗ ./_output/local/bin/darwin/amd64/kube-proxy -v 1 --logging-format json  
I0509 22:10:49.284902   65090 server.go:235] "Warning, all flags other than --config, --write-config-to, and --cleanup are deprecated, please begin using a config file ASAP"
I0509 22:10:49.285015   65090 server_others.go:75] "Using iptables proxy"
I0509 22:10:49.285094   65090 log.go:194] {json 5s 1 [] {{false {{{0 0} {<nil>} 0 DecimalSI}}}}}
E0509 22:10:49.285133   65090 run.go:74] "command failed" err="initialize logging: format: Invalid value: \"json\": Unsupported log format"
```

Comment (cyclinder) [reply cmd/kube-proxy/app/server.go]: json logformat registered here: https://github.com/kubernetes/kubernetes/blob/e865b30abd0a862697ff1d2526ea8897ee98a523/staging/src/k8s.io/component-base/logs/json/register/register.go#L26

Comment (pohly) [reply cmd/kube-proxy/app/server.go]: You need to activate that code by importing the package:
```patch
diff --git a/cmd/kube-proxy/proxy.go b/cmd/kube-proxy/proxy.go
index 86ef9dff68e..085cc2e9f39 100644
--- a/cmd/kube-proxy/proxy.go
+++ b/cmd/kube-proxy/proxy.go
@@ -20,6 +20,7 @@ import (
        "os"
 
        "k8s.io/component-base/cli"
+       _ "k8s.io/component-base/logs/json/register"            // for JSON support
        _ "k8s.io/component-base/metrics/prometheus/restclient" // for client metric registration
        _ "k8s.io/component-base/metrics/prometheus/version"    // for version metric registration
        "k8s.io/kubernetes/cmd/kube-proxy/app"
```

Comment (cyclinder) [reply cmd/kube-proxy/app/server.go]: Thanks! it works now 😄

File: pkg/proxy/apis/config/scheme/testdata/KubeProxyConfiguration/after/v1alpha1.yaml
Comment (pohly) [reply pkg/proxy/apis/config/scheme/testdata/KubeProxyConfiguration/after/v1alpha1.yaml:41]: @liggitt: regarding your proposed type... I think I had suggested something like that and one of the concerns at the time was that tools which write the config will then do with the new encoding (string), while existing components will only support int. Should `MarshalJSON` perhaps encode as int to address that?

@cyclinder: can you perhaps cherry-pick that commit into this PR? Then we can comment on the code directly.


Wasn't one of the concerns about changing the default encoding of that field from int to string that

Comment (cyclinder) [reply pkg/proxy/apis/config/scheme/testdata/KubeProxyConfiguration/after/v1alpha1.yaml:41]: @pohly sure, but cherry-pick that commit into this PR or open another PR for this, which is better?

File: cmd/kube-proxy/app/server.go
Comment (pohly) [inline cmd/kube-proxy/app/server.go:207]: Aren't all command line flags already deprecated in favor of the config fields?

```console
$ go run ./cmd/kube-proxy --logging-format=json
I0510 15:26:19.269444   67867 server.go:235] "Warning, all flags other than --config, --write-config-to, and --cleanup are deprecated, please begin using a config file ASAP"
I0510 15:26:19.269491   67867 server_others.go:75] "Using iptables proxy"
{"ts":1683725179269.8264,"caller":"app/server.go:538","msg":"Neither kubeconfig file nor master URL was specified, falling back to in-cluster config","v":0}
{"ts":1683725179269.8413,"caller":"app/server.go:491","msg":"Error running ProxyServer","err":"unable to load in-cluster configuration, KUBERNETES_SERVICE_HOST and KUBERNETES_SERVICE_PORT must be defined"}
{"ts":1683725179269.864,"caller":"cli/run.go:74","msg":"command failed","err":"unable to load in-cluster configuration, KUBERNETES_SERVICE_HOST and KUBERNETES_SERVICE_PORT must be defined"}
exit status 1
```

With that in mind, does it make sense to register these flags?

Comment (pohly) [reply cmd/kube-proxy/app/server.go:207]: Another observation: ideally, the code which runs before `ValidateAndApplyAsField` shouldn't log anything because the desired logging configuration is still unknown. That's why we have a mixture of text and JSON log output here.

File: pkg/proxy/apis/config/scheme/testdata/KubeProxyConfiguration/after/v1alpha1.yaml
Comment (pohly) [reply pkg/proxy/apis/config/scheme/testdata/KubeProxyConfiguration/after/v1alpha1.yaml:41]: The effect is easier to see when cherry-picked into this PR, so I prefer doing that even if it is, strictly speaking, an unrelated change.

Keep it as a separate commit and we can still merge it separately later.

File: cmd/kube-proxy/app/server.go
Comment (danwinship) [reply cmd/kube-proxy/app/server.go:207]: We've been kind of terrible about actually deprecating the command-line options in favor of config files, but yes, you're right, we should say that if people want to use the cool new features, they have to use the config file

Comment (pohly) [reply cmd/kube-proxy/app/server.go:207]: On the other hand, these flags are the same that also other commands have. For the sake of consistency it would be nice if they were supported also in kube-proxy, not just the config options.

But then the warning needs to be modified so that it doesn't trigger for the logging parameters.

Comment (aojea) [reply cmd/kube-proxy/app/server.go:207]: ref https://github.com/kubernetes/kubernetes/pull/101273#issuecomment-823357862

Comment (cyclinder) [reply cmd/kube-proxy/app/server.go:207]: In the latest changes, I kept the logging-format flag and modified the warning, making sure the code which runs before ValidateAndApplyAsField shouldn't log anything.

Please review the changes, Thanks.

File: cmd/kube-proxy/app/server_others.go
Comment (pohly) [inline cmd/kube-proxy/app/server_others.go:75]: nit: Log messages are meant to start with a capital letter.

File: cmd/kube-proxy/app/server.go
Comment (cyclinder) [reply cmd/kube-proxy/app/server.go:233]: Thanks for the review @pohly, Updated.

File: cmd/kube-proxy/app/server.go
Comment (pohly) [inline cmd/kube-proxy/app/server.go:150]: So now we are back at the point where new flags gets added by the PR (`--logging-format`, etc.) which trigger a deprecation warning:

```
$ go run ./cmd/kube-proxy/ --logging-format=json
Flag --logging-format has been deprecated, This parameter should be set via the config file specified by the Kube-proxy's --config flag.
{"ts":1684218173834.7542,"caller":"app/server_others.go:75","msg":"Using iptables proxy","v":0}
{"ts":1684218173834.9937,"caller":"app/server.go:554","msg":"Neither kubeconfig file nor master URL was specified, falling back to in-cluster config","v":0}
{"ts":1684218173835.0076,"caller":"app/server.go:507","msg":"Error running ProxyServer","err":"unable to load in-cluster configuration, KUBERNETES_SERVICE_HOST and KUBERNETES_SERVICE_PORT must be defined"}
{"ts":1684218173835.024,"caller":"cli/run.go:74","msg":"command failed","err":"unable to load in-cluster configuration, KUBERNETES_SERVICE_HOST and KUBERNETES_SERVICE_PORT must be defined"}
exit status 1
```

I think we have two options:
1. Not add the logging flags to kube-proxy, only configure logging through the config file.
2. Add the flags without marking them as deprecated.

I prefer option 2 because it's consistent with other commands, but I would also be okay with option 1.

Comment (cyclinder) [reply cmd/kube-proxy/app/server.go:150]: I also think the option2 is better, but I see kubelet also marked logging flag as deprecated, it's bug here?

```shell
➜  kubernetes git:(loggin_format) go run ./cmd/kubelet/kubelet.go --logging-format json 
Flag --logging-format has been deprecated, This parameter should be set via the config file specified by the Kubelet's --config flag. See https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/ for more information.
{"ts":1684218864145.937,"caller":"cli/run.go:74","msg":"command failed","err":"failed to construct kubelet dependencies: error reading /var/lib/kubelet/pki/kubelet.key, certificate and key must be supplied as a pair"}
exit status 1

```

Comment (pohly) [reply cmd/kube-proxy/app/server.go:150]: I suppose it wasn't discussed and/or considered at that time.

Comment (cyclinder) [reply cmd/kube-proxy/app/server.go]: So kubelet also needs to be updated?

Comment (pohly) [reply cmd/kube-proxy/app/server.go]: Let's first reach consensus on which of the two options are preferred.

This might be a topic for SIG Architecture or SIG CLI.

@liggitt ?

Comment (liggitt) [reply cmd/kube-proxy/app/server.go]: for components which support a config file that has ~parity with all their CLI flags AND has forward-compatibility guarantees (beta or stable), the flags have generally been marked deprecated in favor of the config-file-specified options

so far, I think that's just kube-scheduler and kubelet

I would not expect kube-proxy to mark CLI flags as deprecated in favor of an alpha-level config file

Comment (cyclinder) [reply cmd/kube-proxy/app/server.go]: yeah, so we don't mark CLI flags as deprecated, Just warn the user that the flag will be deprecated and use the config file instead? Just like the current PR changes, Do you like it？

Comment (liggitt) [reply cmd/kube-proxy/app/server.go]: we shouldn't warn about use of flags in favor of alpha-level config files, only once those graduate to versions with forward compatibility guarantees

Comment (cyclinder) [reply cmd/kube-proxy/app/server.go]: thanks, updated.

File: cmd/kube-proxy/app/server.go
Comment (liggitt) [reply cmd/kube-proxy/app/server.go:233]: separate commit/PR seems fine, but we shouldn't be warning people to stop using flags if the config file is still alpha level

Comment (aojea) [reply cmd/kube-proxy/app/server.go:233]: @liggitt  we have this situation for years in kube-proxy and we are not proud of it, but the reality is that there is no better alternative than use the config, the level is just because nobody spend time in graduating it

Comment (liggitt) [reply cmd/kube-proxy/app/server.go:233]: > no better alternative than use the config

I mean... they could use the flags we're warning them to stop using :)

Comment (aojea) [reply cmd/kube-proxy/app/server.go:233]: we had issues because config override flags and people get confused

Comment (liggitt) [reply cmd/kube-proxy/app/server.go:233]: warning in that case seems fine; here we're warning if they *didn't* specify a config file

Comment (cyclinder) [reply cmd/kube-proxy/app/server.go:233]: So we go back to this warning, and I open an issue for tracking this. see https://github.com/kubernetes/kubernetes/issues/118090

Comment (cyclinder) [reply cmd/kube-proxy/app/server.go:233]: But if we go back to this warning,  the  log warning runs before ValidateAndApplyAsField, Is it acceptable?

**Comment by danwinship** [comment]
@pohly @liggitt is this lgtm from your perspective now?

**Comment by pohly** [comment]
The [discussion around `flushFrequency`](https://github.com/kubernetes/kubernetes/pull/117800#discussion_r1188584225) is still open.

@cyclinder: can you cherry-pick https://github.com/liggitt/kubernetes/commits/flushType as a separate commit and then we can comment on it here? But beware that Jordan is out this month, so we'll not get much feedback from him for a while.

**Comment by cyclinder** [comment]
@pohly Thanks, I missed this and it has now been updated.

File: staging/src/k8s.io/component-base/logs/api/v1/types.go
Comment (pohly) [inline staging/src/k8s.io/component-base/logs/api/v1/types.go:67]: I suspect that encoding as string is a potentially breaking change:
some tool might use this API to generate configuration files for Kubernetes 1.27. Once that tool gets updated, it writes a config that 1.27 cannot parse.

This is completely hypothetical. I am not aware of any such tool.

A safer alternative would be to encode as integer. As long as no-one ever develops a tool that writes configs, it won't make a difference. But if someone does, it'll be annoying that the result is not as human-friendly as a string.

@liggitt ?

Comment (liggitt) [reply staging/src/k8s.io/component-base/logs/api/v1/types.go:67]: I'd tend to favor the clear / human-friendly option... if there was such a tool it could use the 1.27 API definition until it no longer needed to support 1.27 schedulers

Comment (cyclinder) [reply staging/src/k8s.io/component-base/logs/api/v1/types.go:67]: So do we need any changes here?

Comment (pohly) [reply staging/src/k8s.io/component-base/logs/api/v1/types.go:67]: No. Let's go with "human friendly" (= current code).

**Comment by cyclinder** [comment]
Hi @liggitt @pohly Can we merge this PR before code-freeze?

File: cmd/kube-proxy/app/server_test.go
Comment (pohly) [inline cmd/kube-proxy/app/server_test.go:243]: nit: Why this extra `cmp.Diff`?

This could also be written as:
```
if diff := cmp.Diff(config, expected); diff != "" {
    t.Fatalf("unexpected config for %s, diff = %s", tc.name, diff)
}
```

**Review by pohly — COMMENTED** [review]
LGTM to me, with one minor nit.

I'll let @liggitt decide whether the duration commit should be a separate PR. His approval will be needed either way.

File: cmd/kube-proxy/app/server_test.go
Comment (cyclinder) [reply cmd/kube-proxy/app/server_test.go:243]: sometimes `DeepEqual` is false but `Diff` is empty, which causes the unit test to fail. Yes, Using `Cmp.Diff` is perhaps simpler and more accurate.

Comment (pohly) [reply cmd/kube-proxy/app/server_test.go]: Can you give an example? Is such a corner case a failure or something that `Diff` doesn't handle (properly)?

Comment (cyclinder) [reply cmd/kube-proxy/app/server_test.go]:
```
git diff
...
-               if cmp.Diff(config, expected) != "" {
+               if !reflect.DeepEqual(expected, config) {
                        t.Fatalf("unexpected config for %s, diff = %s", tc.name, cmp.Diff(config, expected))
                }
        }
go test -v  ./cmd/kube-proxy/app/ -run TestLoadConfig 
=== RUN   TestLoadConfig
    server_test.go:244: unexpected config for iptables mode, IPv4 all-zeros bind address, diff = 
--- FAIL: TestLoadConfig (0.00s)
=== RUN   TestLoadConfigFailures
=== RUN   TestLoadConfigFailures/Decode_error_test
=== RUN   TestLoadConfigFailures/Bad_config_type_test
=== RUN   TestLoadConfigFailures/Missing_quotes_around_::_bindAddress
--- PASS: TestLoadConfigFailures (0.00s)
    --- PASS: TestLoadConfigFailures/Decode_error_test (0.00s)
    --- PASS: TestLoadConfigFailures/Bad_config_type_test (0.00s)
    --- PASS: TestLoadConfigFailures/Missing_quotes_around_::_bindAddress (0.00s)
FAIL
FAIL    k8s.io/kubernetes/cmd/kube-proxy/app    0.950s
FAIL

```

Comment (pohly) [reply cmd/kube-proxy/app/server_test.go]: Why does `reflect.DeepEqual` return false here? Is that a bug in `DeepEqual`? If yes, it should be fixed. If no, then the test should continue to use `DeepEqual` and we need to investigate why `cmp.Diff` doesn't report the difference.

**Comment by pohly** [comment]
> LGTM to me, with one minor nit.

I take that back: I am worried that https://github.com/kubernetes/kubernetes/pull/117800#discussion_r1255385175 needs to be resolved first.

File: cmd/kube-proxy/app/server_test.go
Comment (cyclinder) [reply cmd/kube-proxy/app/server_test.go]: yes, I'm investigating what's going on here

Comment (cyclinder) [reply cmd/kube-proxy/app/server_test.go]: I found the reason, some default values are missing in `expected.Logging`,  see https://github.com/kubernetes/kubernetes/compare/3c35b0dddfb28538686b8961e23a98e954f6070e..18a5dd572463f4df7e62196f47f5c5cc7cd691e7

Comment (pohly) [reply cmd/kube-proxy/app/server_test.go]: So `DeepEqual` failed because `expected.Logging.Options.JSON.InfoBufferSize.String` mutates the quantity and that led to a difference that was caught only be `DeepEqual`? Why not by `cmp.Diff`? Ah, probably because `cmp.Diff` supports calling `Quantity.Equal`.

I think using only `cmp.Diff` is the right solution here. It clearly does a better job than `DeepEqual` and the code becomes simpler:
```
if diff := cmp.Diff(config, expected); diff != "" {
    t.Fatalf("unexpected config for %s, diff = %s", tc.name, diff)
}
```

Let's not call it twice, BTW.

With that change, you can revert https://github.com/kubernetes/kubernetes/compare/3c35b0dddfb28538686b8961e23a98e954f6070e..18a5dd572463f4df7e62196f47f5c5cc7cd691e7 because we don't have to work around the `DeepEqual` limitation.

Comment (cyclinder) [reply cmd/kube-proxy/app/server_test.go]: `cmp.Diff` is supports calling `Quantity.Equal`? 🤔

OK, let's revert these changes...

Comment (pohly) [reply cmd/kube-proxy/app/server_test.go]: I supports calling *any* Equal method, not `Quantity.Equal` specifically.

**Review by pohly — APPROVED** [review]
/lgtm

File: staging/src/k8s.io/component-base/logs/api/v1/types.go
Comment (liggitt) [inline staging/src/k8s.io/component-base/logs/api/v1/types.go:70]: is this logic tested anywhere?

File: staging/src/k8s.io/component-base/logs/api/v1/types.go
Comment (liggitt) [inline staging/src/k8s.io/component-base/logs/api/v1/types.go:64]: might be worth a comment that this is present only for backwards compatibility for the flushFrequency field, and new fields should use metav1.Duration

**Comment by liggitt** [comment]
unmarshal behavior needs a unit test, then lgtm

**Comment by cyclinder** [comment]
Also, Does the duration commit need to be a separate PR?

File: staging/src/k8s.io/component-base/logs/api/v1/types.go
Comment (cyclinder) [reply staging/src/k8s.io/component-base/logs/api/v1/types.go:66]: Done.

File: staging/src/k8s.io/component-base/logs/api/v1/types.go
Comment (cyclinder) [reply staging/src/k8s.io/component-base/logs/api/v1/types.go:72]: I added some tests for it, see the latest changes.

File: pkg/proxy/apis/config/scheme/testdata/KubeProxyConfiguration/after/v1alpha1.yaml
Comment (aojea) [inline pkg/proxy/apis/config/scheme/testdata/KubeProxyConfiguration/after/v1alpha1.yaml:46]: This is defaulting the verbosity to 0 and overriding the flag value, that was the common way to configure the verbosity of kube-proxy, you can see per example that our CI no longer leverages the flag verbosity https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-kind-e2e-parallel/1689121775398424576/artifacts/kind-control-plane/pods/kube-system_kube-proxy-qz9f6_8e49f584-22b5-4008-9442-b2fae2a3ea75/kube-proxy/0.log

**Pull request merged**
